### DISCUSSION TITLE
IOS: Does Immich download high-resolution photos/videos before saving?

### DISCUSSION THREAD

**Discussion by KatyPurry** [immich-app/immich#4267] [Q&A]
Hello,
My photo library contains 30,000 photos for about 120 GB. On my iPhone I've selected "[Optimized Storage](https://support.apple.com/en-us/HT205703)", which means I only have a few GB used on my iPhone (the thumbnails), with the full format files on iCloud. If I run an Immich backup from my phone, does it fetch the high-resolution files from iCloud or only back up the thumbnails on my phone? 

Thank you

**Comment by ioslife** [comment]
response from Alex https://github.com/immich-app/immich/discussions/1682#discussioncomment-7276130

**Comment by ioslife** [comment]
I am running a backup now. I went and chose a random photo from 2022 that was uploaded to Immich. The iCloud photo and the Immich photo were both the same filesize (2.4mb), filetype(heic), and had the same metadata. The only difference was the Immich version didn't have the Live Photo attached to it.

**Reply by ioslife** [reply]
Of note that _some_ photos on Immich do have the Live Photo though. Need to look into it a bit more to see if there is any rhyme or reason.

**Reply by uninvitedguest** [reply]
I'm curious to know the results when you perform a mass backup instead of a single test. It's everything uploaded to Immich the full file?

Further, what happens on the phone during this process? Does the local storage fill up and need clearing or does it handle the iCloud download, Immich upload, then deletion of the locally stored higher quality image?

**Reply by ioslife** [reply]
@uninvitedguest, all of my photos and videos have now backed up. My iCloud Photos storage is currently sitting at 254.4GB. In Immich, I selected to upload all of my folders, including iCloud Share Libraries. ~~My folder size on Immich is 275GB.~~

~~Downloading photos from Immich and the Photos app on my Mac provides the same filesize as well.~~

Downloading _some_ photos from Immich and the Photos app on my Mac provides the same filesize as well.

My iPhone storage never increased and stuff didn't need to be deleted.

One of the devs would have to speak to how this works exactly, but it apparently worked right?

**Reply by ioslife** [reply]
Ahhh I have to correct the above info:

The 275GB Immich folder includes the thumbnails and encoded videos that Immich has done. My actual photo Library is sitting at 212GB. That is a full 32GB smaller than what is showing on iCloud, so I have to assume that some of the images are not the full quality. Maybe this is because of some images not storing as a Live photo? @alextran1502, any thoughts on that?

I think the next option would be to try icloudpd and uploading that to Immich? Not sure how that would play with backing up photos from the device though. Do photos get overwritten? Is Immich intelligent enough to detect duplicate photos and not upload it if it is already in the Immich Library?

**Reply by alextran1502** [reply]
I cannot tell about the 32GB smaller.

The mobile app will upload everything that is not on the server. The server will then detect if the file is already exists by the file content (checksum) and reject it. The mobile app will then put the file into the black list to not upload it again the next time.

**Reply by ioslife** [reply]
Your response times are amazing hahaha. Thanks! I'll see about trying out the icloudpd method then. Gotta make sure all the photos are original quality. 😄

**Reply by alextran1502** [reply]
No problem, I practically live here

**Comment by ioslife** [comment]
Doing a test today with the following setup:
1. run icloudpd to download full resolution photos onto my NAS
2. run immich bulk upload CLI
3. sync immich iOS app to see if duplicate photos are handled as expected (high-res stays, low-res deleted)

Will report back with findings.

**Reply by ioslife** [reply]
I have a feeling that duplicate photos will not be handled as one would expect. https://github.com/immich-app/immich/issues/2567 talks about why.

**Reply by ioslife** [reply]
Alright, I have some findings to report. I'm not really sure what to say conclusively, but I will attach some images and whatnot for clarity.

For this test, I created a small photo album with six (6) images. I created a brand new user in Immich,  I then used icloudpd to download the photos, then ran Immich bulk upload via CLI, finally logged into the new account on my iPhone and added the test album to be the only album to upload.

The results were not what I expected based on #2567. 

Web:  
<img width="1710" alt="Screenshot 2023-10-18 at 1 08 53 PM" src="https://github.com/immich-app/immich/assets/5478660/a4a81ed9-2ad7-4f56-b8e8-2265c7af4d8d">

iOS:
![IMG_0489](https://github.com/immich-app/immich/assets/5478660/9490cd72-5d31-41b8-9f81-daf7161d7448)


As you can see in the screenshots above, Immich recognized five (5) of the photos as "Asset was uploaded from this device and is now backed up in the cloud/server and still available in original on the device." 

One of the images showed up as "Asset is only available in the cloud and was uploaded from some other device (like the web client) or was deleted from this device after upload."

and that same image also showed up as "Asset is only available locally and has not yet been backed up."

The one image that is acting strange is fascinating because it uses the new iPhone 15 auto-portrait mode setting. The version that shows as available in the cloud is the image _without_ the portrait filter enabled and the one that shows as local only is the image _with_ the portrait filter enabled.

When manually tapping the "Upload Asset" button, the iOS app claims to have successfully uploaded the photo, but it never truly uploads.

Six photos is a very small sample size, but this test seems to have worked as I would hope. I'll consider running this test against my entire iCloud library, but I need to install a new HDD into my NAS first lol. 

@alextran1502, just wanted to tag you into this to see if you had any thoughts about the experiment or any insights into why the one photo is acting up.

**Reply by alextran1502** [reply]
>The one image that is acting strange is fascinating because it uses the new iPhone 15 auto-portrait mode setting. The version that shows as available in the cloud is the image without the portrait filter enabled and the one that shows as local only is the image with the portrait filter enabled.

Hmm, so if I understand correctly, this is the sight for me to buy the new iPhone to troubleshoot this.

**Reply by ioslife** [reply]
> Hmm, so if I understand correctly, this is the sight for me to buy the new iPhone to troubleshoot this.

Lol, you're welcome

**Reply by jrasm91** [reply]
This can happen sometime when the files do not share the same hash. Can you compute the sha1 for each of them?

**Reply by ioslife** [reply]
@jrasm91 
Non-portrait: 252b98e9c7241ffc0f6a2f6ea97554f7f0bea4ef
Portrait: f8a51e9a03ec1aa5d57f6b3d2f7e9700fd00a3e1

**Reply by ioslife** [reply]
Made more space on my server. Testing with entire iCloud Library.

**Reply by ioslife** [reply]
A few hours into the sync with the iOS app. It is going slow since unfortuantely the app has to upload the photos/videos before checking if the photo exists already. 

That said, I am feeling pretty good about what is happening so far. 2,000 photos/videos have uploaded from my phone and only 110 Immich has marked as new unique assets. I'm not sure which of the assets it has marked like that yet and I suspect I won't know until the full sync has happened.

**Reply by ioslife** [reply]
Update: My app keeps crashing while uploading. It gets to around 2800 photos left, crashes, and then when I re-open the app it says 5700 pics again.

Are there logs anywhere I can send to help debug @alextran1502 ?

**Reply by ioslife** [reply]
These are the only errors in the logs I see. 

2023-10-22 20:40:52.844984 | LogLevel.SEVERE | ImmichErrorLogger | Catch all error: PathNotFoundException: Cannot open file, path = '/private/var/mobile/Containers/Data/Application/9DC5D147-DE13-4D07-A953-7F9B2FDD46B1/tmp/.image/B7AFDE55-4A94-4943-82D1-DCBC29C8B6FA_L0_001_0.000000_o_IMG_0257.JPG' (OS Error: No such file or directory, errno = 2) - PathNotFoundException: Cannot open file, path = '/private/var/mobile/Containers/Data/Application/9DC5D147-DE13-4D07-A953-7F9B2FDD46B1/tmp/.image/B7AFDE55-4A94-4943-82D1-DCBC29C8B6FA_L0_001_0.000000_o_IMG_0257.JPG' (OS Error: No such file or directory, errno = 2) | #0      _checkForErrorResponse (dart:io/common.dart:55)#1      _File.open.<anonymous closure> (dart:io/file_impl.dart:381)<asynchronous suspension>#2      _FileStream._start.onOpenFile (dart:io/file_impl.dart:153)<asynchronous suspension>
-- | -- | -- | -- | --

**Reply by ioslife** [reply]
2023-10-22 20:12:27.135324 | LogLevel.SEVERE | ImmichErrorLogger | Catch all error: PathNotFoundException: Cannot delete file, path = '/private/var/mobile/Containers/Data/Application/9DC5D147-DE13-4D07-A953-7F9B2FDD46B1/tmp/.image/432613E3-823B-4692-94F5-989FED385497_L0_001_1695609414.773195_o_57460614191__F24146EA-C418-479C-90AB-105DC3FF05DE.JPG' (OS Error: No such file or directory, errno = 2) - PathNotFoundException: Cannot delete file, path = '/private/var/mobile/Containers/Data/Application/9DC5D147-DE13-4D07-A953-7F9B2FDD46B1/tmp/.image/432613E3-823B-4692-94F5-989FED385497_L0_001_1695609414.773195_o_57460614191__F24146EA-C418-479C-90AB-105DC3FF05DE.JPG' (OS Error: No such file or directory, errno = 2) | #0      _File.throwIfError (dart:io/file_impl.dart:675)#1      _File._deleteSync (dart:io/file_impl.dart:329)#2      FileSystemEntity.deleteSync (dart:io/file_system_entity.dart:424)#3      BackupService.backupAsset (package:immich_mobile/modules/backup/services/backup.service.dart:337)<asynchronous suspension>#4      BackupNotifier.startBackupProcess (package:immich_mobile/modules/backup/providers/backup.provider.dart:461)<asynchronous suspension>#5      BackupNotifier._resumeBackup (package:immich_mobile/modules/backup/providers/backup.provider.dart:591)<asynchronous suspension>
-- | -- | -- | -- | --
2023-10-22 20:03:58.560536 | LogLevel.SEVERE | ImmichErrorLogger | Catch all error: PathNotFoundException: Cannot open file, path = '/private/var/mobile/Containers/Data/Application/9DC5D147-DE13-4D07-A953-7F9B2FDD46B1/tmp/.video/C9F09C1B-AE4E-46C4-BDF5-470EFF4941E6_L0_001_1697845324.201544_o_IMG_2298.MOV' (OS Error: No such file or directory, errno = 2) - PathNotFoundException: Cannot open file, path = '/private/var/mobile/Containers/Data/Application/9DC5D147-DE13-4D07-A953-7F9B2FDD46B1/tmp/.video/C9F09C1B-AE4E-46C4-BDF5-470EFF4941E6_L0_001_1697845324.201544_o_IMG_2298.MOV' (OS Error: No such file or directory, errno = 2) | #0      _checkForErrorResponse (dart:io/common.dart:55)#1      _File.open.<anonymous closure> (dart:io/file_impl.dart:381)<asynchronous suspension>#2      _FileStream._start.onOpenFile (dart:io/file_impl.dart:153)<asynchronous suspension>

**Reply by alextran1502** [reply]
I suspect this will have something to do with iCloud and assets in the shared album feature 🤔

**Reply by ioslife** [reply]
This time I only selected “Recents” as the backup album. There shouldn’t be any shared photos there as far as I know

**Comment by ioslife** [comment]
Update: the iOS backup has finished! It uploaded 1675 unique assets to my Immich server. 

I'd say this is definitely the best way to get your iCloud Library into Immich. There were a few hiccups along the way, but now I have my entire iCloud Photo Library in Immich, at full resolution, with all Live Photos as they should be.

**Reply by ioslife** [reply]
Tagging for notification: @uninvitedguest @2600box

**Reply by dodg3r** [reply]
@ioslife. So how did you do this. Through immich ios app or did you first uploaded through immic bulk upload?
Whats your suggestion for me :) 
I´m about to upload 150 000 photos from my icloud.

**Reply by ioslife** [reply]
@dodg3r, Best way to do it from all my tests above is to use icloudpd to download your iCloud photos/videos in full resolution. Then use Immich CLI to upload to your Immich instance. After that, sync your iPhone to Immich.

**Reply by maffinca69** [reply]
@ioslife Hello. I tried uploading all my photos and videos from icloud to immich (from CLI). Everything works correctly, but how to synchronize all photos and videos with my iPhone? Right now it has all the same files on it as on the server, but after uploading N files, the upload just stops and goes no further (hangs 100%). Maybe I need to completely clear the gallery on my iPhone? 
Ideally I would like to do this so that all photos are stored on the server and not on the iPhone. That way I save space on my phone
P.S. I uploaded 15.000 files

**Reply by ioslife** [reply]
@maffinca69, I think this happened to me. I just had to keep stopping and starting the upload again.

**Reply by daschillinger** [reply]
Can you keep the ios albums and import them to immich? I tried the icloudpd and downloaded only 1 album but got year and month structure. How does immich know what's an album?

**Reply by undaunt** [reply]
@ioslife Did you download both the live photos and static photos from icloudpd and were they somehow put back 'together' in Immich?

**Reply by daschillinger** [reply]
In this single album there were mov and heic files for the Live Photos but I did not upload yet. I am just curious how immich knows what to put together to an album. Are the albums preserved?

**Reply by ioslife** [reply]
Yes!

**Reply by daschillinger** [reply]
What command do I use? Just upload the whole zip? Or refer to the unzipped folder?
Right now I'm downloading with the default parameters with icloudpd. I hope that's right for getting everything and the albums.

**Reply by ioslife** [reply]
I think you have to do the unzipped, but I don’t recall.

**Reply by daschillinger** [reply]
this is all I got
![image](https://github.com/immich-app/immich/assets/155530266/637c1ce3-555e-40fd-8a16-b661d0af7e7c)

**Reply by daschillinger** [reply]
I can't see any album. maybe you did something different to see albums with names? maybe the regular zip download from the icloud page?

**Reply by ioslife** [reply]
I don't think icloudpd exports the albums by default

**Reply by ioslife** [reply]
@daschillinger sorry, my "yes" response earlier was to @undaunt

**Reply by daschillinger** [reply]
Oh, it took the whole day 🤕

**Reply by undaunt** [reply]
@ioslife One issue I noticed with this is there is an upstream limitation in either pyicloud or icloudpd where it can only download the original format of photos, not anything edited with an on-device sidecar. This means portrait mode photos will download without portrait, and any other edits made with cropping or adjustments on an iPhone or MacOS Photos will not be reflected in the icloudpd download. I wound up having about 2k photos out of 30k get re-uploaded from my phone after doing the CLI load, and those photos are now "duplicated" in my library. (Not actually duplicated, because all the portrait photos are half with bokeh, and half without, but you take my point.)

**Reply by ioslife** [reply]
Oh that's fascinating. I didn't realize that happened

**Reply by undaunt** [reply]
@ioslife This is possibly being fixed soon. In the interim I'm also testing steilerDev's icloud-photos-sync.

**Reply by alcroito** [reply]
Did anyone figure out how to preserve the photo albums when exporting using icloudpd and importing into immich?

**Reply by almehdi45** [reply]
@ioslife I'm a bit late to the party here, but was wondering if you could share the details of your workflow. You said that you downloaded it first using icloudpd and then immich bulk upload does this mean you are storing the pictures 2 times ? Do you continue to use iCloudpd or was it only for the first upload ?

**Comment by daschillinger** [comment]
I exported every single album and added it as a folder in the library folder with the option to read folders as albums. Am 14.11.2024 um 10:57 schrieb alcroito ***@***.***>:﻿
Did anyone figure out how to preserve the photo albums when exporting using icloudpd and importing into immich?
